### PR TITLE
Price estimation Fee API

### PR DESCRIPTION
**Pull request by W3stside**
## Summary
- Creates mock api endpoint fn, `getFee` for grabbing FeeInformation object, passing sellToken address as arg

### CODE REVIEW

**Comment by gnosis-info** [comment]
Travis automatic deployment:
   * **🎩 [Dapp](https://pr36--gpv2swap.review.gnosisdev.com)**: Testing web app

File: src/custom/state/fee/hooks.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  token: string // token address
+}
+
+type AddFeeCallback = (addTokenParams: AddFeeParams) => void
+type ClearFeeCallback = (clearTokenParams: ClearFeeParams) => void
+
+export const useFee = (tokenAddress?: string): FeeInformation | undefined => {
+  const { feesMap } = useSelector<AppState, FeeInformationState>(state => state.fee)
+
+  return tokenAddress ? feesMap[tokenAddress]?.fee : undefined
```
Comment (Velenir) [inline src/custom/state/fee/hooks.ts:20]: Put the whole thing inside `useSelector`
react-redux does clever strict equals in store listener and rerenders only if needed
Currently will rerender for you if anything in `feesMap` changes,doesn't have to

File: src/custom/state/fee/hooks.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+export const useAddFee = (): AddFeeCallback => {
+  const dispatch = useDispatch<AppDispatch>()
+  return (addTokenParams: AddFeeParams) => dispatch(updateFee(addTokenParams))
```
Comment (Velenir) [inline src/custom/state/fee/hooks.ts:25]: `useCallback`

File: src/custom/state/fee/hooks.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+export const useClearFee = (): ClearFeeCallback => {
+  const dispatch = useDispatch<AppDispatch>()
+  return (clearTokenParams: ClearFeeParams) => dispatch(clearFee(clearTokenParams))
```
Comment (Velenir) [inline src/custom/state/fee/hooks.ts:30]: `useCallback`

File: src/custom/state/fee/middleware.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+const isReplaceState = isAnyOf(replaceSwapState)
+
+export const applyFeeMiddleware: Middleware = ({ dispatch, getState }) => next => async action => {
+  if (isCurrencyChangeAction(action)) {
+    const {
+      swap: {
+        OUTPUT: { currencyId }
+      },
+      fee: { feesMap }
+    }: { fee: FeeInformationState; swap: SwapState } = getState()
```
Comment (Velenir) [inline src/custom/state/fee/middleware.ts:19]: Maybe would look better with less destructuring,or destructure in steps

File: src/custom/state/fee/middleware.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    if (currencyIdFromPayload) {
+      const tokenFee = feesMap[currencyIdFromPayload]?.fee
+
+      if (!tokenFee) {
+        const fee = await getFee(currencyIdFromPayload)
+        dispatch(
+          updateFee({
+            token: currencyIdFromPayload,
+            fee
+          })
```
Comment (Velenir) [inline src/custom/state/fee/middleware.ts:37]: So it only updates if not already fetched before? What about stale fees?

File: src/custom/state/fee/reducer.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+// {token address => FeeInformationObject} mapping
+type FeesMap = Record<string, FeeInformationObject>
+
+export interface FeeInformationState {
+  readonly feesMap: Partial<FeesMap>
+}
+
+const initialState: FeeInformationState = {
+  feesMap: {}
```
Comment (Velenir) [inline src/custom/state/fee/reducer.ts:23]: Should we also differentiate by `chainId` or do we count on token addresses not overlapping over diff networks?

File: src/custom/utils/fees.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+ */
+export const getFee = async (currencyId: string): Promise<FeeInformation> => {
+  return new Promise(accept => {
+    return setTimeout(() => {
+      console.debug(
+        `[MOCK] utils/price::getFee ==> https://protocol.dev.gnosisdev.com/api/v1/fee/${currencyId}`,
+        MOCK_FEE_INFORMATION
+      )
+      accept(MOCK_FEE_INFORMATION)
+    }, Math.floor(Math.random() * 1000))
```
Comment (Velenir) [inline src/custom/utils/fees.ts:33]: We already have delay in `custom/utils/misc`, added by @anxolin

**Review by Velenir — APPROVED** [review]

File: src/custom/utils/fees.ts (hunk 7 above)
Comment (W3stside) [reply src/custom/utils/fees.ts:33]: @Velenir yeah added to the other PR that watersfalls in, should have been in this one i guess

File: src/custom/state/fee/hooks.ts (hunk 1 above)
Comment (W3stside) [reply src/custom/state/fee/hooks.ts]: cool thanks

File: src/custom/state/fee/middleware.ts (hunk 5 above)
Comment (W3stside) [reply src/custom/state/fee/middleware.ts:43]: good point, updated

File: src/custom/state/fee/reducer.ts (hunk 6 above)
Comment (W3stside) [reply src/custom/state/fee/reducer.ts:23]: hmm true, likely will. @anxolin thoughts

File: src/custom/state/fee/middleware.ts (hunk 4 above)
Comment (W3stside) [reply src/custom/state/fee/middleware.ts]: ok, done

File: src/custom/state/fee/reducer.ts (hunk 6 above)
Comment (anxolin) [reply src/custom/state/fee/reducer.ts:23]: token addresses do overlap in different networks

**Review by Velenir — APPROVED** [review]

**Pull request merged**